### ISSUE TITLE
Add GPG agent/GPG support

### ISSUE THREAD

**Issue by strarsis** [dlech/KeeAgent#173]
It would be great if KeeAgent could also be used with GPG keys as GPG agent.

**Comment by dlech** [comment]
Yes, I will have to learn how this works.

Do you use Linux, Mac or Windows? I don't know of a GPG agent program for Windows, so this may just be a Linux-only feature.

**Comment by strarsis** [comment]
My usecase would be letting it run as agent on windows and access it from 
inside a Linux VM (similar to using a socket from inside Babun/Cygwin.

**Comment by nifr** [comment]
GPG agent support would be soooo nice. Can I assist in gathering necessary information somehow @dlech ? :)

**Comment by dlech** [comment]
A link to a GPG agent protocol specification would be a good start.

**Comment by strarsis** [comment]
https://www.gnupg.org/documentation/manuals/gnupg/Agent-Protocol.html

**Comment by markschuh** [comment]
gpg-agent has "--enable-putty-support" since gnupg-2.0.21, so it can be used as pageant replacement on windows. see https://lists.gnupg.org/pipermail/gnupg-announce/2013q3/000332.html and http://www.bootc.net/archives/2013/06/09/my-perfect-gnupg-ssh-agent-setup/
Hence keeagent can already communicate with gpg-agent, when started in client or auto agent mode.

But I'm unsure yet about the use cases, where such a setup would be that useful. gpg-agent is the preferred use method, if the secrets are stored inside a hardware device - which is more secure than the software wallet keepass. So if I already use gpg-agent and store the private keys on hardware - what would be the advantage to bring keeagent (or even keepass) into that setup?

I have tried the combination - and at the moment the only visible difference is  - with respect to pure keeagent use in main agent mode -  that I have to add a *new* passphrase to each single of my keepass saved ssh-keys, when keeagent (in client mode) ask gpg-agent to store them temporarily.
(gpg-agent fires pine-entry for passphrase entry, when asked to store a new secret by keeagent)

**Comment by r2evans** [comment]
@markschuh  I was just looking into this; I suspect that *"it's not that easy"* because, unless I'm misunderstanding things, KeeAgent is *emulating* `ssh-agent`, not calling it. If the latter were the case, then it might suffice to replace the call to `ssh-agent` with `gpg-agent` (even on windows, perhaps). But that is predicated on the called-process having access to the keys stored within the keepass kdbx, and I'm going to guess (@dlech  let me know if there's something I'm not aware of) that implementing a virtual filesystem for the purpose of the called process is non-trivial. (Having said that, if that could be done, then it might reduce the complexity of using dependent protocols like this.)

@dlech, I'm assuming this is still a nice-to-have but not yet in your immediate plans. Besides oodles of money and working around a day job, is there something that somebody else can do to assist?

**Comment by kevinvalk** [comment]
I am interested for this on Windows as well (enigmail).

**Comment by dlech** [comment]
does Windows/Enigmail use a pinentry type program as well?

**Comment by r2evans** [comment]
I'm just starting to incorporate GPG into my thought process and workflow, so I cannot inform on the best or most popular method. How can I tell? I'll do what I can (short of opening RDP on my laptop, sorry).

If it changes anything, the percentage use-case for me will be commit-signing (`git`), not email/enigmail. (Unfortunately, most of my clients are comfortable with S/MIME, while the others are unaware/uncaring.)

**Comment by dlech** [comment]
> enigmail

I think that KeeFox would be a better fit for this use case. I didn't see anything in the Enigmail docs that mentioned the use of a GPG agent.

**Comment by r2evans** [comment]
Except that KeeFox will not work with future versions of TB. I don't exactly when.I see your point, though, so perhaps this primarily/only helps the use of signed commits in git (a potentially high-frequency use).On Mar 24, 2018 12:44, David Lechner <notifications@github.com> wrote:
enigmail

I think that KeeFox would be a better fit for this use case. I didn't see anything in the Enigmail docs that mentioned the use of a GPG agent.

**Comment by ctd1500** [comment]
> does Windows use a pinentry type program as well?

Yes, the [windows gpg port](https://www.gpg4win.org/features.html) has pinentry, as do other frontends.

**Comment by lusitania** [comment]
> (@dlech) I'm assuming that people wanting this feature are using Linux?

I'm on Windows +KeePass +KeeAgent +Cygwin, linked over socket ($SSH_AUTH_SOCK).

**Comment by mendhak** [comment]
Both Windows (Win10, KeePass,KeeAgent,git bash) as well as Linux (Ubuntu, KeePass, KeeAgent).

Similar to a few others here, main use case is for git commits, but also package signing.

**Comment by typoworx-de** [comment]
I would really love such a solution. I'm on Ubuntu/Linux with Keepass & Thunderbird+Enigmail and it is really annoying entering the password manually by copy-paste from KeePass.

I did some discovery on how pinentry works, which is the default way on linux presenting any kind of PGP-Prompt for Password. It implements GPG-Agent-Protocol. I found a very basic implementation of it for Bash!
https://gist.github.com/barn/e3ff508c3032da3ff905

For research I implemented some Logging to figure out what commands are invoked for a PGP-PIN Session. Cmd and Rest represent the variables read from StdIn (see line 24 of in the link to the bash script).
``
Cmd: OPTION
Rest: no-grab
Cmd: OPTION
Rest: allow-external-password-cache
Cmd: OPTION
Rest: default-ok=_OK
Cmd: OPTION
Rest: default-cancel=_Cancel
Cmd: OPTION
Rest: default-yes=_Yes
Cmd: OPTION
Rest: default-no=_No
Cmd: OPTION
Rest: default-prompt=PIN:
Cmd: OPTION
Rest: default-pwmngr=_Save in password manager
Cmd: OPTION
Rest: default-cf-visi=Do you really want to make your passphrase visible on the screen?
Cmd: OPTION
Rest: default-tt-visi=Make passphrase visible
Cmd: OPTION
Rest: default-tt-hide=Hide passphrase
Cmd: OPTION
Rest: touch-file=/home/test-user/.gnupg/S.gpg-agent
Cmd: GETINFO
Rest: pid
Cmd: SETKEYINFO
Rest: --clear
Cmd: SETDESC
Rest: Please enter the passphrase to unlock the OpenPGP secret key:%0A%22Max Mustermann <max@mustermann.com>%22%0A4096-bit RSA key, ID AXAXAX,%0Acreated 2018-12-27.%0A
Cmd: SETPROMPT
Rest: Passphrase:
Cmd: GETPIN
Rest: 
Cmd: BYE
Rest:
``

The GETPIN CMD then simply responses the plain Password! It should be very easy to implement this into keepass or wrap the script around kpcli (keepass command-line agent for linux). Before doing this myself I am in hope there's already something in work for KeePass?

**Comment by djherbis** [comment]
@typoworx-de Because of this issue I actually went ahead and implemented a (super minimal, but functional) plugin for myself which does this.

I only implemented a couple of the pinentry commands (most of them you can just respond OK to and not do anything). The only real important one was GETPIN for me.

Basic impl:
- Create a binary which binds stdin/stdout to a TCP connection, set this binary as your gpg-agent pinentry program.
- Create a KeePass Plugin which runs a TCP server, and handle the pinentry commands (basically just GETPIN, also did a couple GETINFO ones, and SETKEYINFO, but they don't really impact my impl. much).

I pretty much hard-coded which keepass entry it uses for the gpg password, and used env vars for sharing the port the server is running on (I also used TLS, but that's not that important for a localhost only server).

If you think this ^ would be handy, or if there's a better way to handle comm between the pinnetry program / keepass, let me know and I'll see if I can release my repo, it's not pretty, but it's been working for me for awhile.

**Comment by jballe** [comment]
> If you think this ^ would be handy, or if there's a better way to handle comm between the pinnetry program / keepass, let me know and I'll see if I can release my repo, it's not pretty, but it's been working for me for awhile.

I think it would at least be interesting to see, maybe it is enough on a personal level. So I will appreciate if you can make it available.

**Comment by djherbis** [comment]
@jballe It's rough, but here you go https://github.com/djherbis/keepass-pinentry

**Comment by elieux** [comment]
My usecase is calling GPG from command line on Windows (from a Cygwin-like environment).

Currently when the private key needs to be used, a GUI pinentry is spawned and I can auto-type into it. The agent (or some related part) then remembers something in that shell session so the password doesn't have to entered repeatedly (for a limited time).

Implementing a pinentry (most probably has to be an executable) that *securely* connects to KeePass and retrieves the password sounds like the easier solution.

Implementing the whole agent would probably enable more control, but sounds more difficult.

**Comment by djherbis** [comment]
@elieux My solution does basically what you describe. A pinentry executable, which communicates with KeePass over an encrypted, localhost TCP connection to get the password.

Are there parts of the security that you're worried about?

**Comment by ilippert** [comment]
A minor comment. I guess given that enigmail is not maintained anymore and thunderbird's gpg compatibility is now moved into thunderbird (and seems to be quite closed off), the usecase keepass+enigmail ceases to be of any relevance anymore.

**Comment by BrainStone** [comment]
I'd have a use for signing git commits

**Comment by r2evans** [comment]
> A minor comment. I guess given that enigmail is not maintained anymore and thunderbird's gpg compatibility is now moved into thunderbird (and seems to be quite closed off), the usecase keepass+enigmail ceases to be of any relevance anymore.

You're assuming that everybody who uses GPG (1) only needs GPG for email, and (2) uses TB. I think both assumptions are incorrect.
